ls1intum/Artemis
# `Programming exercises`: Allow test weights to be all zero if manual feedback is enabled

### Checklist
#### General



#### Server

#### Client

#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on ~~Test Server 2~~ the local setup (Jenkins and Gitlab).
    - The change affects programming exercises, but is independent of the used Git/CI-system. Therefore, a check on TS1 + local should be enough.

### Motivation and Context
We use the automatic tests purely as guideline to students on their progress in the exercise. The actual scoring, however, is only determined by manual feedback.
In the current implementation programming exercises with manual feedback need to have at least one test case that gives points to students. If only automatic feedback is used for the scoring, this makes sense as otherwise all students could only ever achieve a score of 0%. When manual feedback is enabled, however, this limitation no longer applies.

### Description
I updated both the server- and client-side checks for the test weights to allow a weight-sum of zero for programming exercises only if manual feedback is enabled.

Additionally removes sinon/chai from `programming-exercise-configure-grading.component.spec.ts`. The test is probably only slow due to the chart components (`CategoryIssuesChartComponent`, `TestCasePassedBuildsChartComponent`). They are only used in a single test case. Refactoring it to another dedicated file is out of scope of this PR, as neither one of the components does have its own test suite yet.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with only automatic feedback

Steps:
1. Log in as the Instructor.
2. Go to the ‘Grading’ page of the exercise management.
3. Change all test weights to zero and try to save.
    - An error message should be shown that this is not valid.
    - Saving should only be possible if all weights are >= 0 and at least one weight is >0.
4. In the exercise configuration, change the assessment type to manual.
5. Repeat step 3.
    - This time no error message should be shown. The weights are saved.
    - Trying to set individual test weights to values <0 should still be invalid.
6. Now try to edit the exercise again back to only automatic assessments.
    - This should not be allowed and an error message should be shown.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage
| File | Instruction | Branch |
| --- | --- | --- |
| `ProgrammingExerciseTestCaseService.java` | 92% | 77% |
| `ProgrammingExerciseResource.java` | 92% | 83% |
| `programming-exercise-configure-grading.component.ts` | 89.9% | 73.7% |
